File: RepoBrain/examples/python-app/src/app/leads/create_lead.py
"""Lead creation workflow."""
from app.common.db import leads_repository
from app.common.logger import get_logger
from app.common.validation import require_non_empty, validate_email
from app.leads.crm_client import send_lead_to_crm
from app.leads.lead import Lead, LeadInput
from app.users.phone import normalize_phone

logger = get_logger(__name__)


def create_lead(input: LeadInput) -> Lead:
    """Validate a lead, persist it and forward it to the external CRM."""
    name = require_non_empty(input.name, "name")
    email = validate_email(input.email)
    phone = normalize_phone(input.phone) if input.phone else None

    payload = {
        "name": name,
        "email": email,
        "phone": phone,
        "source": input.source,
        "message": input.message,
    }
    response = send_lead_to_crm(payload)

    lead = Lead(
        id=0,
        name=name,
        email=email,
        phone=phone,
        source=input.source,
        remote_id=response.remote_id,
    )
    lead_id = leads_repository.insert(lead)
    lead = lead.model_copy(update={"id": lead_id})
    logger.info("Created lead %s from source %s", lead_id, input.source)
    return lead

File: RepoBrain/examples/typescript-app/src/modules/leads/createLead.ts
/**
 * Lead creation use case. Validates the input, persists the lead, and hands it
 * off to the CRM. This is the function wired to `POST /leads`.
 */

import { generateId, getRepository } from "../../common/db.js";
import { logger } from "../../common/logger.js";
import { sendLeadToCrm } from "../crm/crm-client.js";
import type { Lead, LeadInput } from "./lead.schema.js";
import { validateLeadPayload } from "./validateLead.js";

const leads = getRepository<Lead>("leads");

/**
 * Creates a lead from an untrusted payload and forwards it to the CRM.
 *
 * @param input Raw lead payload (already shaped as `LeadInput` by the caller,
 *   but re-validated here to be safe).
 */
export async function createLead(input: LeadInput): Promise<Lead> {
  const validated = validateLeadPayload(input);

  const lead: Lead = {
    id: generateId("lead"),
    name: validated.name,
    email: validated.email,
    phone: validated.phone,
    source: validated.source ?? "web",
    status: "new",
    createdAt: new Date().toISOString(),
  };

  await leads.insert(lead);
  logger.info("lead created", { leadId: lead.id, source: lead.source });

  try {
    const crmId = await sendLeadToCrm(lead);
    lead.crmId = crmId;
    lead.status = "sent_to_crm";
    await leads.insert(lead);
    logger.info("lead synced to crm", { leadId: lead.id, crmId });
  } catch (error) {
    lead.status = "failed";
    await leads.insert(lead);
    logger.error("lead crm sync failed", {
      leadId: lead.id,
      reason: error instanceof Error ? error.message : String(error),
    });
  }

  return lead;
}
